flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -0,0 +1,237 @@
+
+        // Create message builder for sending message.
+        TypedMessageBuilder<IN> builder = createMessageBuilder(topic, element, sinkContextAdapter);
+
+        // Perform message sending.
+        if (deliveryGuarantee == DeliveryGuarantee.NONE) {
+            // We would just ignore the sending exception. This may cause data loss.
+            builder.sendAsync();
+        } else {
+            // Waiting for permits to write message.
+            pendingMessages.acquire();
+            CompletableFuture<MessageId> sender = builder.sendAsync();
+            sender.whenComplete(
+                    (id, ex) -> {
+                        pendingMessages.release();
+                        if (ex != null) {
+                            mailboxExecutor.execute(
+                                    () -> {
+                                        throw new FlinkRuntimeException(
+                                                "Failed to send data to Pulsar " + topic, ex);
+                                    },
+                                    "Failed to send message to Pulsar");
+                        } else {
+                            LOG.debug("Sent message to Pulsar {} with message id {}", topic, id);
+                        }
+                    });
+        }
+    }
+
+    private TypedMessageBuilder<IN> createMessageBuilder(
+            String topic, IN element, PulsarSinkContextAdapter context) {
+        TypedMessageBuilder<IN> builder =
+                producerRegister.createMessageBuilder(topic, element, schema);
+
+        byte[] orderingKey = serializationSchema.orderingKey(element, context);
+        if (orderingKey != null) {
+            builder.orderingKey(orderingKey);
+        }
+
+        String key = serializationSchema.key(element, context);
+        if (!Strings.isNullOrEmpty(key)) {
+            builder.key(key);
+        }
+
+        Map<String, String> properties = serializationSchema.properties(element, context);
+        if (properties != null) {
+            builder.properties(properties);
+        }
+
+        Long timestamp = serializationSchema.eventTime(element, context);
+        if (timestamp != null) {
+            builder.eventTime(timestamp);
+        }
+
+        List<String> clusters = serializationSchema.replicationClusters(element, context);
+        if (clusters != null) {
+            builder.replicationClusters(clusters);
+        }
+
+        if (serializationSchema.disableReplication(element, context)) {
+            builder.disableReplication();
+        }
+
+        return builder;
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException {
+        if (endOfInput) {
+            // Try flush only once when we meet the end of the input.
+            producerRegister.flush();
+        } else {
+            while (pendingMessages.availablePermits() < sinkConfiguration.getMaxPendingMessages()) {
```

Comment:
Let me explain this in detail.

`pendingMessages` is created by using `new Semaphore(sinkConfiguration.getMaxPendingMessages())`, the available permits would only be equal to or less than the maximum. We would block the message writing by `pendingMessages.acquire()` until we finally have a permit to write a message.

We would trigger `flush(false)` before doing the checkpoint. The `pendingMessages.availablePermits() == sinkConfiguration.getMaxPendingMessages()` means that there is on pending message. Because we would return permits when the message writing is complete in the mailbox thread by `pendingMessages.release()`.

I think the confusion for you is just that the `Semaphore` is not commonly used in Java. XD